File: apps/go/internal/advent_of_code_2023/numbers.go
package adventofcode2023

// Define the custom type
type Digit int

// Define the constants for each digit
const (
	One Digit = iota + 1
	Two
	Three
	Four
	Five
	Six
	Seven
	Eight
	Nine
)

// Implement the methods for the custom type
func (d Digit) AsString() string {
	switch d {
	case One:
		return "one"
	case Two:
		return "two"
	case Three:
		return "three"
	case Four:
		return "four"
	case Five:
		return "five"
	case Six:
		return "six"
	case Seven:
		return "seven"
	case Eight:
		return "eight"
	case Nine:
		return "nine"
	default:
		return "unknown"
	}
}

func (d Digit) AsInt() int {
	return int(d)
}

func (d Digit) Len() int {
	return len(d.AsString())
}

File: apps/rust/advent_of_code_2023/src/numbers.rs
pub enum Number {
    One,
    Two,
    Three,
    Four,
    Five,
    Six,
    Seven,
    Eight,
    Nine,
    Zero,
}
impl Number {
    pub fn as_str(&self) -> &str {
        match self {
            Number::One => "one",
            Number::Two => "two",
            Number::Three => "three",
            Number::Four => "four",
            Number::Five => "five",
            Number::Six => "six",
            Number::Seven => "seven",
            Number::Eight => "eight",
            Number::Nine => "nine",
            Number::Zero => "zero",
        }
    }
    pub fn as_int(&self) -> u8 {
        match self {
            Number::One => 1,
            Number::Two => 2,
            Number::Three => 3,
            Number::Four => 4,
            Number::Five => 5,
            Number::Six => 6,
            Number::Seven => 7,
            Number::Eight => 8,
            Number::Nine => 9,
            Number::Zero => 0,
        }
    }
    pub fn length(&self) -> usize {
        self.as_str().len()
    }
}
